### PR TITLE
Add scope

### PR DESCRIPTION
**Pull request by ybelMekk**
This feature will give NAIS application more autonomy and transparency when configuring and exposing  scopes to other organizations.

### CODE REVIEW

**Comment by ybelMekk** [comment]
Missing metrics for scope do.

**Review by tronghn — COMMENTED** [review]
looks fairly good, noted some points for discussion.

File: controllers/common/reconciler.go [hunk 1]
```diff
@@ -390,3 +405,195 @@ func (r *Reconciler) updateInstance(ctx context.Context, instance clients.Instan
 
 	return updateFunc(existing)
 }
+
+func scopesExist(tx Transaction, exposedScopes map[string]naisiov1.ExposedScope) (*scopes.ScopeStash, error) {
```
[the remark this answers is not in the document]
Comment (ybelMekk) [reply controllers/common/reconciler.go:409]: agree.. had hard time keeping up, but it makes seens.

File: pkg/digdir/types/types.go [hunk 2]
```diff
@@ -89,3 +92,55 @@ type Scope struct {
...
+	VisibilityInternal Visibility = "INTERNAL"
+	VisibilityPrivate  Visibility = "PRIVATE"
+	VisibilityPublic   Visibility = "PUBLIC"
+)
+
+type TokenType string
+
+const (
+	TokenTypeSelfContained TokenType = "SELF_CONTAINED"
+	VisibilityOpaque       TokenType = "OPAQUE"
```
[the remark this answers is not in the document]
Comment (ybelMekk) [reply pkg/digdir/types/types.go:108]: done

File: pkg/digdir/types/types.go [hunk 3]
```diff
@@ -89,3 +92,55 @@ type Scope struct {
...
 	State         ScopeAccessState `json:"state"`
 }
+
+type Visibility string
+
+const (
+	VisibilityInternal Visibility = "INTERNAL"
+	VisibilityPrivate  Visibility = "PRIVATE"
+	VisibilityPublic   Visibility = "PUBLIC"
+)
```
[the remark this answers is not in the document]
Comment (ybelMekk) [reply pkg/digdir/types/types.go:102]: done

File: pkg/digdir/types/types.go [hunk 4]
```diff
@@ -89,3 +92,55 @@ type Scope struct {
...
+	Visibility                 Visibility `json:"visibility"`
+}
+
+type State string
+
+const (
+	StateRequested State = "REQUESTED"
+	StateApproved  State = "APPROVED"
+	StateDenied    State = "DENIED"
+)
```
[the remark this answers is not in the document]
Comment (ybelMekk) [reply pkg/digdir/types/types.go:137]: done

File: controllers/common/reconciler.go [hunk 5]
```diff
@@ -215,6 +214,22 @@ func (r *Reconciler) process(tx *Transaction) error {
...
+
+		if exposedScopes != nil {
+			filteredScopes, err := scopesExist(*tx, exposedScopes)
+			if err != nil {
+				return fmt.Errorf("checking if scopes exists: %w", err)
+			}
+
+			if err := r.handleFilteredScopes(filteredScopes, tx, exposedScopes); err != nil {
+				return fmt.Errorf("handle filtered scopes: %w", err)
+			}
```
[the remark this answers is not in the document]
Comment (ybelMekk) [reply controllers/common/reconciler.go:229]: done

File: controllers/common/reconciler.go [hunk 6]
```diff
@@ -390,3 +405,195 @@ func (r *Reconciler) updateInstance(ctx context.Context, instance clients.Instan
...
+func (r *Reconciler) CreateScope(tx *Transaction, instance *naisiov1.MaskinportenClient, newScope naisiov1.ExposedScope) (*types.ScopeRegistration, error) {
+	scopeRegistrationPayload := clients.ToScopeRegistration(instance, newScope)
+	scope, err := r.createScope(tx, scopeRegistrationPayload)
+	if err != nil {
+		return nil, fmt.Errorf("creating scope: %w", err)
+	}
+	return scope, nil
+}
+
+func (r *Reconciler) InActivateScope(tx *Transaction, scope string) (*types.ScopeRegistration, error) {
```
[the remark this answers is not in the document]
Comment (ybelMekk) [reply controllers/common/reconciler.go:586]: done

File: pkg/digdir/client.go [hunk 7]
```diff
@@ -160,3 +167,99 @@ func clientMatches(actual types.ClientRegistration, desired clients.Instance) bo
...
+	endpoint := fmt.Sprintf("%s/scopes/access/%s?scope=%s", c.Config.DigDir.Admin.BaseURL, consumerOrgno, url.QueryEscape(scope))
+	registration := &types.ConsumerRegistration{}
+
+	if err := c.request(ctx, http.MethodPut, endpoint, nil, registration); err != nil {
+		return nil, fmt.Errorf("updating scope acl: %w", err)
+	}
+	return registration, nil
+}
+
+func (c Client) InActivateConsumer(ctx context.Context, scope, consumerOrgno string) (*types.ConsumerRegistration, error) {
```
[the remark this answers is not in the document]
Comment (ybelMekk) [reply pkg/digdir/client.go:257]: done

File: config/samples/maskinportenclient.yaml [hunk 8]
```diff
@@ -3,8 +3,16 @@ apiVersion: nais.io/v1
...
   scopes:
-    - name: "nav:test/api"
+    exposes:
+      # nav:arbeid:digdirator.test.read
+      - name: "digdirator.test.read"
+        enabled: true
+        product: "arbeid"
+        consumers:
+          - orgno: "889640782"
+    use:
```
[the remark this answers is not in the document]
Comment (ybelMekk) [reply config/samples/maskinportenclient.yaml:17]: hard, both is good but with one is better?

File: controllers/common/reconciler.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ybelMekk) [reply controllers/common/reconciler.go]: done

File: config/samples/maskinportenclient.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (ybelMekk) [reply config/samples/maskinportenclient.yaml:17]: i make it consumes

File: pkg/clients/instance_test.go [hunk 9]
```diff
@@ -44,31 +45,51 @@ func TestGetSecretJwkKey(t *testing.T) {
...
-		assert.True(t, actual)
+		assert.Equal(t, expected, actual)
 	})
 
 	t.Run("IDPortenClient with changed value should not be up-to-date", func(t *testing.T) {
 		client := minimalIDPortenClient()
+		expected := client.GetStatus().GetSynchronizationHash()
 		client.Spec.ClientURI = "changed"
-		actual, err := clients.IsUpToDate(client)
+		actual, err := calculateHash(client)
```
Comment (tronghn) [inline pkg/clients/instance_test.go:58]: kan byttes ut med `client.Hash()`

File: README.md [hunk 10]
```diff
@@ -28,8 +28,43 @@ An example of resources is available in [config/samples/idportenclient.yaml](con
...
+---
+apiVersion: v1
+kind: Namespace
+metadata:
+  managedFields:
+    - apiVersion: v1
+      manager: kube-apiserver
+      operation: Update
+  name: test
+```
```
[the remark this answers is not in the document]
Comment (ybelMekk) [reply README.md]: done

File: pkg/clients/instance_test.go (hunk 9 above)
Comment (ybelMekk) [reply pkg/clients/instance_test.go]: done

**Review by tronghn — APPROVED** [review]

**Review by tronghn — APPROVED** [review]

**Pull request merged**